crates/block-producer/src/server.rs

```diff
@@ -211,8 +212,27 @@ impl BlockProducerRpcServer {
     }
 
     async fn serve(self, listener: TcpListener) -> Result<(), tonic::transport::Error> {
+        // Configure the trace layer with callbacks.
+        let trace_layer = TraceLayer::new_for_http()
+            .make_span_with(miden_node_utils::tracing::grpc::block_producer_trace_fn)
+            .on_request(|request: &http::Request<_>, _span: &Span| {
+                info!(
+                    "request: {} {} {} {:?}",
+                    request.method(),
+                    request.uri().host().unwrap_or("unknown_host"),
+                    request.uri().path(),
+                    request.headers()
+                );
+            })
+            .on_response(|response: &http::Response<_>, latency: Duration, _span: &Span| {
+                info!("response: {} {:?}", response.status(), latency);
+            })
+            .on_failure(|error: ServerErrorsFailureClass, latency: Duration, _span: &Span| {
+                error!("error: {} {:?}", error, latency);
+            });
```

Comment:
@Mirko-von-Leipzig I have checked this locally and on honeycomb. LMK any changes you would make here.
![Screenshot 2025-02-22 at 10 45 51 AM](https://github.com/user-attachments/assets/582a10c0-bc01-4f2d-a73b-1751eae3498a)

Reply:
Right now you're emitting events, but I'd like to attach whatever we can as attributes on the span.

For context, the open-telemetry specification has a bunch of (experimental) suggestions:

- https://opentelemetry.io/docs/specs/semconv/rpc/rpc-spans/#server-attributes
- https://opentelemetry.io/docs/specs/semconv/rpc/grpc/

We have access to only some of them here I think.

Reply:
I just can't find a way to surface lower level details like client addr to the relevant stacks here.

The `tonic::Request` struct has `::remote_addr() -> Option<SocketAddr>` but we can't get that type wired through into our trace stack here AFAICT, only `http::Request`. And I can't find a way to get headers to contain the remote IP address via `http::Request` in any way.

Relevant links for reference
https://github.com/hyperium/tonic/issues/430
https://github.com/tower-rs/tower-http/blob/main/examples/tonic-key-value-store/src/main.rs#L191

Reply:
Should be able to get socket addr via this when building the tonic server:
https://docs.rs/tonic/0.2.1/tonic/transport/server/trait.Connected.html

Will give it a go

Reply:
Wow this was crazy hard to find: https://github.com/tower-rs/tower-http/discussions/428.

Only by chance did I go check in the discussions - no other google/search-kungfu found it.

I believe this is what we need? It may also mean that we have to just do our own layer to extract some of the info before handling the routes/request/responses.

Reply:
Ah, well done! This looks to have worked. e.g.:
```rust
.on_request(|request: &http::Request<_>, _span: &Span| {
    info!(
        "request: {} {} {} {} {:?}",
        request
            .extensions()
            .get::<tonic::transport::server::TcpConnectInfo>() // as per axum example above
            .unwrap()
            .remote_addr()
            .unwrap(),
```

I'll put together all the changes from your comments and put this in review next.